### DISCUSSION TITLE
Can relationship `load_only` attributes be set at the mapper level?

### DISCUSSION THREAD

**Discussion by LonelyVikingMichael** [sqlalchemy/sqlalchemy#10398] [Usage Questions] [answered]
I often make use of `load_only()` at the query level to only load required columns for a particular relationship. Consider the following:

```python
class Child(Base):
    name: Mapped[str] = mapped_column(...)
    other_column: Mapped[...] = mapped_column(...)

    
class Parent(Base):
    name: Mapped[str] = mapped_column(...)
    
    _children: Mapped[list[Child]] = relationship(lazy="noload")
    
    @property
    def child_names(self) -> list[str]:
        """The names of the associated child objects."""
        return [child.name for child in self._children]
```
In this example, `Parent` has no need to access `Child.other_column`, we're only interested in accessing the `child_names` property.

We can write the select statement like this:
`select(Parent).options(selectinload(Parent.children).load_only(Child.name))`

I'd like to know if it is possible to set this "globally" at the mapper level instead? Just to be clear, I'm not looking to defer loading of `Child.other_column` entirely - only within context of `Parent.children`.

**Comment by CaselIT** [comment]
Hi,

No I don't think there is anything that directly support that.
There is are defer groups that kinda do the opposite of what you want to achieve https://docs.sqlalchemy.org/en/20/orm/queryguide/columns.html#loading-deferred-columns-in-groups

So basically you could configure `Child` to load only the attribute you need when used in a relationship with all other columns deferred and part of the same group, then use `undefer_group` to load all attribute when loading child directly.

@zzzeek mike would the following make sense: extend how `deferred_group` works:
- if `deferred=True` in the column the current behaviour is in place, so you use `undefer_group` to load these attributes
- if `deferred=False` in the column a new behaviour is in place (IIRC this currently raises): the attribute is usually loaded, but you can defer it using a new `defer_group`

**Comment by zzzeek** [comment, marked as answer]
hi -

OK, just read your last line, since yes I was giving you the usual answer to use https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.mapped_column.params.deferred.

There's no mapper-level config that would apply a particular set of loader options in a nested fashion like that, I'm not sure what such a mapping API would look like, it would be something like being able to send loader options into `relationship()` directly which would have a lot of implications / complexity / etc.

So otherwise the only way to get this to be "automatic" would be using the [do_orm_execute()](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.do_orm_execute) event hook to detect this condition.

to do this in a fully generalized way that supports an eager load at any level of depth requires a bit of inside information.  Here is a POC

```py
from __future__ import annotations

from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import ForeignKey
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Load
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import selectinload
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class Child(Base):
    __tablename__ = "child"

    id: Mapped[int] = mapped_column(primary_key=True)

    name: Mapped[str]
    other_column: Mapped[int]
    parent_id: Mapped[int] = mapped_column(ForeignKey("parent.id"))


class Parent(Base):
    __tablename__ = "parent"

    id: Mapped[int] = mapped_column(primary_key=True)

    name: Mapped[str]

    _children: Mapped[list[Child]] = relationship(lazy="noload")

    @property
    def child_names(self) -> list[str]:
        """The names of the associated child objects."""
        return [child.name for child in self._children]


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)


@event.listens_for(Session, "do_orm_execute")
def receive_do_orm_execute(orm_execute_state):
    if (
        orm_execute_state.is_relationship_load
        and orm_execute_state.loader_strategy_path[-1]
        is Parent._children.property
    ):
        # simple way - just make a loader option and assume the
        # original query was from Parent.  Won't work for deeper eager loads.
        # opt = defaultload(Parent._children).load_only(Child.name)

        # fancy way - make a Load object that's relative to the full
        # eager load path.  requires PathRegistry manipulation to construct
        # this path
        opt = Load._construct_for_existing_path(
            orm_execute_state.loader_strategy_path[Child.__mapper__]
        ).load_only(Child.name)

        return orm_execute_state.invoke_statement(
            orm_execute_state.statement.options(opt)
        )


s = Session(e)

s.add(
    Parent(
        name="p1",
        _children=[Child(name=f"c{i}", other_column=i) for i in range(3)],
    )
)
s.commit()


stmt = select(Parent).options(selectinload(Parent._children))


r = s.scalars(stmt).all()
```

**Reply by LonelyVikingMichael** [reply]
Thanks Mike, I'll give the "fancy way" a try. I really appreciate the response.

**Reply by LonelyVikingMichael** [reply]
Hi Mike

It looks like this condition works a charm when using `selectinload()`:
```python
if (
    orm_execute_state.is_relationship_load
    and orm_execute_state.loader_strategy_path[-1]
    is Parent._children.property
):
    ...
```
But is not true for `joinedload()`. Is there a way to apply the same loader criteria to joined eager loads?

**Reply by zzzeek** [reply]
the joinedload() is not a separate SQL query so you'd have to intercept it within the main query, that is, when orm_execute_state.is_relationship_load is False.  then you'd have to dig into the query to figure out that there's a joinedload option with the path that you're looking for.      

maybe just try adding `defaultload(Parent._children).load_only(Child.name)` to all queries and see if it complains.
